dask_saturn/core.py

```diff
@@ -1,92 +1,78 @@
 
 
 class SaturnCluster(SpecCluster):
+    cluster_url = None
+    _scheduler_address = None
+    _dashboard_link = None
+    _instances = weakref.WeakSet()
+
     def __init__(
         self,
         *args,
-        n_workers=None,
-        cluster_url=None,
+        n_workers=0,
         worker_size=None,
         scheduler_size=None,
-        nprocs=None,
-        nthreads=None,
+        nprocs=1,
+        nthreads=1,
         scheduler_service_wait_timeout=DEFAULT_WAIT_TIMEOUT_SECONDS,
+        close_when_done=False,
+        loop=None,
+        asynchronous=False,
         **kwargs,
     ):
-        if cluster_url is None:
-            self._start(
-                n_workers=n_workers,
-                worker_size=worker_size,
-                scheduler_size=scheduler_size,
-                nprocs=nprocs,
-                nthreads=nthreads,
-                scheduler_service_wait_timeout=scheduler_service_wait_timeout,
-            )
-        else:
-            self.cluster_url = cluster_url if cluster_url.endswith("/") else cluster_url + "/"
-        info = self._get_info()
-        self._dashboard_link = info["dashboard_link"]
-        self._scheduler_address = info["scheduler_address"]
-        self.loop = None
+        self.n_workers = n_workers
+        self.worker_size = worker_size
+        self.scheduler_size = scheduler_size
+        self.nprocs = nprocs
+        self.nthreads = nthreads
+        self.scheduler_service_wait_timeout = scheduler_service_wait_timeout
+        self._loop_runner = LoopRunner(loop=loop, asynchronous=asynchronous)
+        self.loop = self._loop_runner.loop
         self.periodic_callbacks = {}
-
-    @classmethod
-    def reset(
-        cls,
-        n_workers=None,
-        worker_size=None,
-        scheduler_size=None,
-        nprocs=None,
-        nthreads=None,
-        scheduler_service_wait_timeout=DEFAULT_WAIT_TIMEOUT_SECONDS,
-    ):
-        """Return a SaturnCluster
-
-        Destroy existing Dask cluster attached to the Jupyter Notebook or
-        Custom Deployment and recreate it with the given configuration.
-        """
-        print(f"Resetting cluster.")
-        url = urljoin(BASE_URL, "api/dask_clusters/reset")
-        cluster_config = {
-            "n_workers": n_workers,
-            "worker_size": worker_size,
-            "scheduler_size": scheduler_size,
-            "nprocs": nprocs,
-            "nthreads": nthreads,
-        }
-        response = requests.post(url, data=json.dumps(cluster_config), headers=HEADERS)
-        if not response.ok:
-            raise ValueError(response.reason)
-        return cls()
-
-    @property
-    def status(self):
-        if self.cluster_url is None:
-            return "closed"
+        self._lock = asyncio.Lock()
+        self._asynchronous = asynchronous
+        self._instances.add(self)
+        self._correct_state_waiting = None
+        self.close_when_done = close_when_done
+        self.status = "created"
```

Comment:
seems strange to me to have a string "status" field. But I'm guessing this comes from the spec?

Reply:
Yes indeedy. `SpecCluster` does something more smart with it, but the concept is the same, that we set the status and then wait for it to change.